astropy/io/fits/column.py

```diff
@@ -1389,19 +1389,21 @@ def __getattr__(self, name):
 
         Implements for example self.units, self.formats, etc.
         """
-
         cname = name[:-1]
         if cname in KEYWORD_ATTRIBUTES and name[-1] == 's':
             attr = []
-            for col in self:
+            for col in self.columns:
                 val = getattr(col, cname)
-                if val is not None:
-                    attr.append(val)
-                else:
-                    attr.append('')
+                attr.append(val if val is not None else '')
             return attr
         raise AttributeError(name)
 
+    def _update_attrs(self, name=None):
+        """Get or update local lists of column attributes"""
+        attrs = [name] if name is not None else KEYWORD_ATTRIBUTES
+        for name in attrs:
+            setattr(self, name + 's', self.__getattr__(name + 's'))
```

Comment:
@embray - Ok, thanks for the detailed answer. I took another look, and it seems that at the `FITS_rec` things are ok. It still seems interesting to cache `names` and `formats` in `ColDefs` because there are accessed several times here (for the `dtype` and `_recformats` properties).
But to simplify, I dropped the previous commit and used lazy properties for  `names` and `formats`, and use the signal to delete the cached version if needed. It seems much cleaner like this.